src/HPLTokenHandler.mo

```diff
@@ -0,0 +1,331 @@
+            case (#ok vid) {
+              map.put(p, {
+                virtualAccountId = vid;
+                var credit = 0;
+              });
+              vid;
+            };
+            case (#err error) {
+              let message = "Opening virtual account problem";
+              journal.push((Time.now(), p, #error(message, error)));
+              freezeTokenHandler(message);
+              throw Error.reject(message);
+            };
+          };
+        };
+      };
+      (ownPrincipal, virtualAccountId);
+    };
+
+    /// query the usable balance
+    public func balance(p : Principal) : ?Nat = Option.map<Info, Nat>(map.get(p), func (info: Info) : Nat = info.credit);
+
+    /// query journal for debug purposes. Returns:
+    /// 1) array of all items in order, starting from the oldest record in journal, but no earlier than "startFrom" if provided
+    /// 2) the index of next upcoming journal log. Use this value as "startFrom" in your next journal query to fetch next entries
+    public func queryJournal(startFrom: ?Nat): ([JournalRecord], Nat) = (
+      Iter.toArray(
+        journal.slice(
+          Int.abs(Int.max(Option.get(startFrom, 0), journal.pushesAmount() - journalSize)),
+          journal.pushesAmount()
+        )
+      ),
+      journal.pushesAmount(),
+    );
+
+    /// retrieve the current freeze state
+    public func isFrozen() : Bool = isFrozen_;
+
+    /// retrieve the sum of all successful consolidations
+    public func consolidatedFunds() : Nat = consolidatedFunds_;
+
+    /// deduct amount from P’s usable balance. Return false if the balance is insufficient.
+    public func debit(p : Principal, amount : Nat) : Bool {
+      if (isFrozen()) {
+        return false;
+      };
+      let ?info = map.get(p) else return false;
+      if (info.credit < amount) return false;
+      info.credit -= amount;
+      journal.push((Time.now(), p, #debited(amount)));
+      totalDebited += amount;
+      assertBalancesIntegrity();
+      true;
+    };
+
+    ///  add amount to P’s usable balance (the credit is created out of thin air)
+    public func credit(p : Principal, amount : Nat) : Bool {
+      if (isFrozen()) {
+        return false;
+      };
+      let ?info = map.get(p) else return false;
+      info.credit += amount;
+      totalCredited += amount;
+      journal.push((Time.now(), p, #credited(amount)));
+      assertBalancesIntegrity();
+      true;
+    };
+
+    /// The handler will query the balance of virtual account, previously opened for P.
+    /// If there will be non-zero amount (deposit), the handler will turn balance to zero and add the deposit to the credit of P. 
+    /// Returns the newly detected deposit and total usable balance if success, otherwise null
+    public func notify(p : Principal) : async* ?(Nat, Nat) {
+      if (isFrozen()) return null;
+      let ?info = map.get(p) else return null;
+      let updateResult = await hpl.setVirtualBalance(info.virtualAccountId, 0);
```

Comment:
A comment like this would help:
```
// We pass through any call Error instead of catching it
```

Reply:
Hmm, in deposit/withdraw we handle call errors and provide #err result, but in sweep in/out we propagate them. I think we should be consistent in this regard. Let's handle them, as we do in IRCR1 token handler, what do you think?